File: src/net/acmicpc/solved/silver/Solution_1406.java
package net.acmicpc.solved.silver;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;

public class Solution_1406 {

	private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

	// 에디터 - S2
	// https://www.acmicpc.net/problem/1406
	public static void main(String[] args) throws IOException {
		StringBuilder sb = new StringBuilder(readLine());
		int n = readInt();

		int pointer = sb.length();
		for (int i = 0; i < n; i++) {
			String[] input = readLine().split(" ");
			String command = input[0];

			switch (command) {
				case "L":
					if (pointer != 0) {
						pointer -= 1;
					}
					break;
				case "D":
					if (pointer < sb.length()) {
						pointer += 1;
					}
					break;
				case "B":
					if (pointer != 0) {
						sb.delete(pointer - 1, pointer);
						pointer -= 1;
					}
					break;
				case "P":
					String c = input[1];
					sb.insert(pointer, c);
					pointer += 1;
					break;
			}
		}
		write(sb);

		close();
	}


	// Libraries
	private static String readLine() throws IOException {
		return br.readLine();
	}

	private static int readInt() throws IOException {
		return Integer.parseInt(br.readLine()
								  .trim());
	}

	private static int[] readIntegers() throws IOException {
		return Arrays.stream(br.readLine()
							   .trim()
							   .split(" "))
					 .mapToInt(Integer::parseInt)
					 .toArray();
	}

	private static <T> void write(T value) throws IOException {
		String s = String.valueOf(value);
		bw.write(s);
		bw.newLine();
	}

	private static void close() throws IOException {
		bw.flush();
		bw.close();
		br.close();
	}

}

File: src/net/acmicpc/solved/silver/Solution_1406.py
import sys

# libs
input = sys.stdin.readline


# 에디터 - S2
# https://www.acmicpc.net/problem/1406

S = input().strip()
N = int(input())
pointer = len(S)  # 안써도 됨.

left = list(S)    # 커서 기준 왼쪽
right = []        # 커서 기준 오른쪽; 스택으로 사용할거기 때문에 뒤집혀 있다고 생각하기

for i in range(N):
    command = input().split()

    # 커서 이동 - 왼쪽
    if command[0] == 'L':
        if left:
            right.append(left.pop())  # 커서 기준, 오른쪽 스택에서 빼서, 왼쪽 스택에 삽입

    # 커서 이동 - 오른쪽
    elif command[0] == 'D':
        if right:
            left.append(right.pop())  # 커서 기준, 왼쪽 스택에서 빼서, 오른쪽에 삽입

    # 삭제 (커서 왼쪽)
    elif command[0] == 'B':
        if left:
            left.pop()  # 왼쪽 스택(커서 기준 좌측)에 문자 제거

    # 삽입 (커서 왼쪽)
    elif command[0] == 'P':
        left.append(command[1])  # 왼쪽 스택(커서 기준 좌측)에 문자 삽입
        

print(''.join(left + right[::-1]))  # right 스택은 뒤집혀 있기 때문에 다시 뒤집기